### DISCUSSION TITLE
Extension to convert column(s) into an UNION data type

### DISCUSSION THREAD

**Discussion by hinxx** [duckdb/duckdb#6549] [General]
I'm trying to workaround the lack of union support in parquet file format. I have data that is defined by a name, a timestamp and a value. Value data type can be integer, float, double, string, array of those (there are no complex types involved). Having this kind of value in parquet, in a single column, is not possible since something like ORC union is not supported.

One idea I have is to have several columns for value, each holding distinct data type + an additional 'tag' column to specify which one actually holds the data for the row at hand (only one column, others have nulls). For the writer side this seems feasible and straightforward.
For the reader side I would like to abstract this workaround from the end user by "converting" those value columns into a single resulting column (of `LogicalTypeId::UNION` ?) and give that to the user. I'm thinking DuckDB extension could be coded to do that.

Does that make sense? Not knowing the internal of DuckDB I'm not sure how/what can be achieved with and extension.

Another approach would be to serialize the original data type (ie. with msgpack) and put that in binary (BLOB) column on writer side. On the reader side the extension would deserialize the BLOB into final value (again `LogicalTypeId::UNION` ?).. A bit more overhead, but I might use a very light serdes that would just add some data type/size identification byte(s) to at the start of the BLOB payload (again I have no complex types to handle, arrays contain primitive types only) .

Going for UNION data type in DuckDB seems logical to me but I have not yet done and trials with the code.

I'm not really sure what an extension can do and how TBH. I did hack around the extension template a bit to see how the `ScalarFunction` is used, but that is all..

**Comment by orzom411** [comment]
I have a question, the results set being returned you want varying data type to be returned in the same column? 

Maybe you can structure then employ struct_extract? 

```
copy( 
SELECT {'x': cast( 1 as int ), 'y':cast( null as varchar ), 'z': cast( null as timestamp )} as d
union all
SELECT {'y':'MyFile', 'x': cast( null as int ), 'z': cast( null as timestamp )} as d
union all
SELECT {'z': cast(  now() as timestamp ), 'y':cast( null as varchar ), 'x': cast( null as int )} as d
) to 'test.parquet' (format parquet);
```

Select coalesce( struct_extract(d,'y'), struct_extract(d,'x'), struct_extract(d,'z') ) as sc, * from read_parquet( 'test.parquet' );

> ┌────────────────────────┬─────────────────────────────────────────────────────┐
> │           sc           │                          d                          │
> │        varchar         │      struct(x varchar, y varchar, z timestamp)      │
> ├────────────────────────┼─────────────────────────────────────────────────────┤
> │ 1                      │ {'x': 1, 'y': NULL, 'z': NULL}                      │
> │ MyFile                 │ {'x': MyFile, 'y': NULL, 'z': NULL}                 │
> │ 2023-03-03 19:09:08.16 │ {'x': 2023-03-03 19:09:08.16, 'y': NULL, 'z': NULL} │
> └────────────────────────┴─────────────────────────────────────────────────────┘
( * Sorry the table format is ugly, I do not know how to format the output nicely. )

Notice the schema:

`Select` * from parquet_schema('test.parquet');`

The rub with this is as a return element would be of varying data types might really throw off a client process, I don't know as I haven't actually tried to do that with nested structures like this.

Also note, above I flipped around the elements and ended up with an artifact (int became a varchar), need to explicit keep the structure the same, e.g.: 

```
copy( 
SELECT {'x': cast( 1 as int ), 'y':cast( null as varchar ), 'z': cast( null as timestamp )} as d
union all
SELECT {'x': cast( null as int ), 'y':'MyFile', 'z': cast( null as timestamp )} as d
union all
SELECT { 'x': cast( null as int ), 'y':cast( null as varchar ), 'z': cast(  now() as timestamp )} as d
) to 'test.parquet' (format parquet);

Select coalesce( struct_extract(d,'y'), struct_extract(d,'x'), struct_extract(d,'z') ) as sc, * from read_parquet( 'test.parquet' );
```

_Results from_: CLI v0.7.2-dev14

**Reply by hinxx** [reply]
Thanks!! This looks like something I could try out.

> Also note, above I flipped around the elements and ended up with an artifact (int became a varchar), need to explicit keep the structure the same, e.g.:

In my case non-duckDB parquet file writers are going to be doing this (C++) where the schema will be fixed, hence I should be getting consistent data layout.

All the data types in the coalesced column now turn to `varchar`, which is understandable. Something I haven't realized till now, I need to think about the consequences of this, though. I believe this is not a duckDB matter (AFAICT pandas/pyarrow can't handle different data type in Series/Array).
